src/app/delegation_backend/src/submit_test.go

```diff
@@ -0,0 +1,253 @@
+package delegation_backend
+
+import (
+  logging "github.com/ipfs/go-log/v2"
+  "testing"
+  "encoding/base64"
+  "encoding/json"
+  "time"
+  "bytes"
+  "net/http/httptest"
+  "io/ioutil"
+  "math/rand"
+)
+
+const TSPG_EXPECTED_1 = `{"block":"zLgvHQzxSh8MWlTjXK+cMA==","created_at":"2021-07-01T16:21:33Z","peer_id":"MLF0jAGTpL84LLerLddNs5M10NCHM+BwNeMxK78+"}`
+const TSPG_EXPECTED_2 = `{"block":"zLgvHQzxSh8MWlTjXK+cMA==","created_at":"2021-07-01T16:21:33Z","peer_id":"MLF0jAGTpL84LLerLddNs5M10NCHM+BwNeMxK78+","snark_work":"Bjtox/3Yu4cT5eVCQz/JQ+P3Ce1JmCIE7N6b1MAa"}`
+
+func mkB64 (s string) *Base64 {
+  r := new(Base64)
+  r.json = []byte("\"" + s + "\"")
+  r.data, _ = base64.StdEncoding.DecodeString(s)
+  return r
+}
+
+func mkB64B (b []byte) *Base64 {
+  r := new(Base64)
+  r.json = []byte("\"" + base64.StdEncoding.EncodeToString(b) + "\"")
+  r.data = b
+  return r
+}
+
+func TestSignPayloadGeneration(t *testing.T){
+  req := new(submitRequestData)
+  req.PeerId = mkB64("MLF0jAGTpL84LLerLddNs5M10NCHM+BwNeMxK78+")
+  req.Block = mkB64("zLgvHQzxSh8MWlTjXK+cMA==")
+  req.CreatedAt, _ = time.Parse(time.RFC3339, "2021-07-01T19:21:33+03:00")
+  json, err := makeSignPayload(req)
+  if err != nil || !bytes.Equal(json, []byte(TSPG_EXPECTED_1)) {
+    t.FailNow()
+  }
+  req.SnarkWork = mkB64("Bjtox/3Yu4cT5eVCQz/JQ+P3Ce1JmCIE7N6b1MAa")
+  json, err = makeSignPayload(req)
+  if err != nil || !bytes.Equal(json, []byte(TSPG_EXPECTED_2)) {
+    t.FailNow()
+  }
+}
+
+func testSubmitH (maxAttempt int, initWl Whitelist) (*ObjectsToSave, *SubmitH, *timeMock) {
+  storage := make(ObjectsToSave)
+  log := logging.Logger("delegation backend test")
+  app := new(App)
+  app.Log = log
+  app.Save = func(objs ObjectsToSave) {
+    for path, value := range(objs) {
+      storage[path] = value
+    }
+  }
+  counter, tm := newTestAttemptCounter(1)
+  app.SubmitCounter = counter
+  app.Now = tm.Now
+  wlMvar := new(WhitelistMVar)
+  wlMvar.Replace(&initWl)
+  app.Whitelist = wlMvar
+  return &storage, app.NewSubmitH(), tm
+}
+
+func (sh *SubmitH) testRequest(body []byte) *httptest.ResponseRecorder {
+  recorder := httptest.NewRecorder()
+  req := httptest.NewRequest("POST", "http://127.0.0.1/v1/submit", bytes.NewReader(body))
```

Comment:
By factoring out the `v1` we could probably upgrade all tests at once? Also, this URL is repeated across 3 different test cases, maybe we should reference it as a single variable?

Reply:
This string is not actually used. Any URL would do. I made a single constant with this URL to avoid duplication.